src/source-windivert.c

```diff
@@ -0,0 +1,844 @@
+/* Copyright (C) 2018 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Jacob Masen-Smith <jacob@evengx.com>
+ *
+ * WinDivert emulation of netfilter_queue functionality to hook into Suricata's
+ * IPS mode. Supported solely on Windows.
+ *
+ */
+
+#include "suricata-common.h"
+#include "suricata.h"
+#include "tm-threads.h"
+
+#include "util-byte.h"
+#include "util-debug.h"
+#include "util-device.h"
+#include "util-error.h"
+#include "util-privs.h"
+#include "util-ioctl.h"
+
+#include "runmodes.h"
+
+#include "queue.h"
+
+#include "source-windivert-prototypes.h"
+#include "source-windivert.h"
+
+#ifdef WINDIVERT
+// clang-format off
+#include <winsock2.h>
+#include <windows.h>
+#include <iptypes.h>
+// clang-format on
+#endif
+
+#ifndef WINDIVERT
+/* Gracefully handle the case where no WinDivert support is compiled in */
+
+TmEcode NoWinDivertSupportExit(ThreadVars *, const void *, void **);
+
+void TmModuleReceiveWinDivertRegister(void)
+{
+    memset(&tmm_modules[TMM_RECEIVEWINDIVERT], 0, sizeof(TmModule));
```

Comment:
instead of these individual memsets, I think we can memset the whole thing at startup? RegisterAllModules looks like a good candidate.

Reply:
Is this a request that I change this for the PR or a note for your reference? Right now, this seems to be the default behavior for other thread modules to zero themselves.

Reply:
Hmm maybe I need more coffee, but I don't see the other modules doing this. I think it would be a cleaner solution to do the whole thing at once, but it's not a blocker. So if you're up for it, go ahead :)

Reply:
Nope, I need more coffee, I was wrong. I based it on `source-nfq.c` but that uses a bunch of manual `NULL` assignments. I extrapolated the `memset` zeroing from that.

There currently doesn't seem to be any full zeroing of the `tmm_modules` allocation.